File: examples-nosql-go-sdk/sqlexamples/AlterTable.go
// Copyright (c) 2020, 2023 Oracle and/or its affiliates.
// Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/
package main
import (
	"fmt"
	"time"

	"github.com/oracle/nosql-go-sdk/nosqldb"
	"github.com/oracle/nosql-go-sdk/nosqldb/auth/iam"
	"github.com/oracle/nosql-go-sdk/nosqldb/common"
)

// Creates a client with the supplied configurations.
// This function encapsulates environmental differences and returns a
// client handle to use for data operations.
func createClient_cloud() (*nosqldb.Client, error) {
	var cfg nosqldb.Config
	// replace the placeholder with your region identifier
	region := "<your_region_identifier>"
	// Replace the value of config file location and the ocid of your compartment
	sp, err := iam.NewSignatureProviderFromFile("<location_config_file>","","","<your_compartment_ocid>")
	if err != nil {
		return nil, fmt.Errorf("cannot create a Signature Provider: %v", err)
	}
	cfg = nosqldb.Config{
	        Mode:                  "cloud",
	     	  Region:                common.Region(region),
			  AuthorizationProvider: sp,
	}
   client, err := nosqldb.NewClient(cfg)
	return client, err
}
// Creates a client with the supplied configurations for onPremise database
func createClient_onPrem() (*nosqldb.Client, error) {
	var cfg nosqldb.Config
	// replace the placeholder with the fullname of your host
	endpoint := "http://<hostname>:8080"
	cfg= nosqldb.Config{
      Endpoint: endpoint,
      Mode:     "onprem",
   }
	// If using a secure store pass the username, password of the store to Config
	// cfg := nosqldb.Config{
   //    Mode:     "onprem",
   //    Username: "<username>",
   //    Password: []byte("<password>"),
   // Specify InsecureSkipVerify
   //    HTTPConfig: httputil.HTTPConfig{
   //        InsecureSkipVerify: true,
   //    },
   client, err := nosqldb.NewClient(cfg)
	return client, err
}
	// Creates a table
func createTable(client *nosqldb.Client, err error, tableName string)(){
	  stmt := fmt.Sprintf("CREATE TABLE IF NOT EXISTS %s ("+
		 "acct_Id INTEGER," +
		 "profile_name STRING," +
		 "account_expiry TIMESTAMP(1) ," +
		 "acct_data JSON, " +
		 "PRIMARY KEY(acct_Id))",tableName)
	  tableReq := &nosqldb.TableRequest{
		    Statement: stmt,
		    TableLimits: &nosqldb.TableLimits{
			     ReadUnits:  20,
			     WriteUnits: 20,
			     StorageGB:  1,
		    },
	  }
	  tableRes, err := client.DoTableRequest(tableReq)
	  if err != nil {
		   fmt.Printf("cannot initiate CREATE TABLE request: %v\n", err)
		   return
	  }
	  // The create table request is asynchronous, wait for table creation to complete.
	  _, err = tableRes.WaitForCompletion(client, 60*time.Second, time.Second)
	  if err != nil {
      fmt.Printf("Error finishing CREATE TABLE request: %v\n", err)
      return
	 }
	 fmt.Println("Created table ", tableName)
	 return
}
// alter an existing table and add a column
func alterTable(client *nosqldb.Client, err error, tableName string)(){
	stmt := fmt.Sprintf("ALTER TABLE %s (ADD acctName STRING)",tableName)
	tableReq := &nosqldb.TableRequest{
		Statement: stmt,
	}
	tableRes, err := client.DoTableRequest(tableReq)
	if err != nil {
		fmt.Printf("cannot initiate ALTER TABLE request: %v\n", err)
		return
	}
	// The alter table request is asynchronous, wait for table alteration to complete.
	_, err = tableRes.WaitForCompletion(client, 60*time.Second, time.Second)
	if err != nil {
		fmt.Printf("Error finishing ALTER TABLE request: %v\n", err)
		return
	}
	fmt.Println("Altered table ", tableName)
	return
}
// drop an existing table
func dropTable(client *nosqldb.Client, err error, tableName string)(){
	stmt := fmt.Sprintf("DROP TABLE %s",tableName)
	tableReq := &nosqldb.TableRequest{
		Statement: stmt,
	}
	tableRes, err := client.DoTableRequest(tableReq)
	if err != nil {
		fmt.Printf("cannot initiate DROP TABLE request: %v\n", err)
		return
	}
	// The drop table request is asynchronous, wait for table drop to complete.
	_, err = tableRes.WaitForCompletion(client, 60*time.Second, time.Second)
	if err != nil {
		fmt.Printf("Error finishing DROP TABLE request: %v\n", err)
		return
	}
	fmt.Println("Dropped table ", tableName)
	return
}

func main() {
	// if using cloud service uncomment the line below. else if using onPremises comment this line out
	client, err := createClient_cloud()
	// if using onPrem uncomment the line below, else if using cloud service, comment this line
   // client, err := createClient_onPrem()
	if err != nil {
		fmt.Printf("cannot create NoSQL client: %v\n", err)
		return
	}
	defer client.Close()
	tableName := "stream_acct"
  	createTable(client, err,tableName)
	alterTable(client, err,tableName)
	dropTable(client, err,tableName)
}

File: examples-nosql-node-sdk/sqlexamples/AlterTable.js
/* Copyright (c) 2023, 2024 Oracle and/or its affiliates.
 * Licensed under the Universal Permissive License v 1.0 as shown at
 * https://oss.oracle.com/licenses/upl/
 */
'use strict';
const NoSQLClient = require('oracle-nosqldb').NoSQLClient;
const TABLE_NAME = 'stream_acct';

/*
 * Call the main function for this example
 */
doaltertable();

async function doaltertable() {
   let handle;
   try {
      /* UNCOMMENT line of code below if you are using Oracle NoSQL Database
      * Cloud service. Leave the line commented if you are using onPrem database
      handle = await getConnection_cloud(); */
      /* UNCOMMENT line of code below if you are using onPremise Oracle NoSQL
       * Database. Leave the line commented if you are using NoSQL Database
       * Cloud Service
      handle = await getConnection_onPrem(); */
      await createTable(handle);
      await alterTable(handle);
      await dropTable(handle);
   } catch (error ) {
      console.log(error);
   }
   finally {
      if (handle) {
         handle.close();
      }
   }
}

/* Create and return an instance of a NoSQLCLient object for cloud service */
function getConnection_cloud() {
   /* replace the placeholders for compartment and region with actual values.*/
   const Region = `<your_region_identifier>`;
   return new NoSQLClient({
      region: Region,
      compartment: "<ocid_of_your_compartment>",
   });
}
/* Create and return an instance of a NoSQLCLient object for onPremises */
function getConnection_onPrem() {
   /* replace the placeholder with the fullname of your host */
   const kvstore_endpoint = `http://<hostname>:8080`;
   return new NoSQLClient({
      serviceType: "KVSTORE",
      endpoint: kvstore_endpoint
   });
   /* if it is a secure store, comment the return statement above and
   * uncomment the lines below and pass the credentials, username and password
   return new NoSQLClient({
      "serviceType": "KVSTORE",
      "endpoint": "",
      "auth": {
         "kvstore":{
            "user": "",
            "password": ""
         }
      }
   }); */
}
/* creates a table */
async function createTable(handle) {
   const createDDL = `CREATE TABLE IF NOT EXISTS ${TABLE_NAME} (acct_Id INTEGER,
                                                                 profile_name STRING,
                                                                 account_expiry TIMESTAMP(1),
                                                                 acct_data JSON,
                                                                 primary key(acct_Id))`;
   console.log('Create table: ' + createDDL);
   let res =  await handle.tableDDL(createDDL, {
            complete: true,
            tableLimits: {
                readUnits: 20,
                writeUnits: 20,
                storageGB: 1
            }
   });
   console.log('Table created: ' + TABLE_NAME);
}
/* alter a table and add a column */
async function alterTable(handle) {
   const alterDDL = `ALTER TABLE ${TABLE_NAME} (ADD acctname STRING)`;
   let res =  await handle.tableDDL(alterDDL);
   console.log('Table altered: ' + TABLE_NAME);
}
/* drop a table */
async function dropTable(handle) {
   const dropDDL = `DROP TABLE ${TABLE_NAME}`;
   let res =  await handle.tableDDL(dropDDL);
   console.log('Table dropped: ' + TABLE_NAME);
}
